### ISSUE TITLE
High retention results in INITIALIZING-shards for a long time

### ISSUE THREAD

**Issue by Freeaqingme** [m3db/m3#1733]
In a single m3dbnode deployment with embedded etcd all shards appear to remain in state INITIALIZING forever.

Config: https://gist.github.com/Freeaqingme/afe68e7bbff6bdd17f8c062c0dd24e21

Machine has 4G of RAM and is meant to store a relative small amount of metrics for a 2 year period, so a few settings were modified:
```
  local:
    namespaces:
      - namespace: default
        type: unaggregated
        retention: 17520h

  cache:
    series:
      policy: lru
    postingsList:
      size: 262

  commitlog:
    flushMaxBytes: 524
    flushEvery: 1s
    queue:
      calculationType: fixed
      size: 2048
```

After starting the instance, the database is created:
```
curl -v -X POST http://localhost:7201/api/v1/database/create -d '{
  "type": "local",
  "namespaceName": "default",
  "retentionTime": "17520h"
}'    
```

Afterwards, the shards are shown as INITIALIZING (shard id 1-62 removed for brevity):
```
curl  localhost:7201/api/v1/placement | python -m json.tool  

100  5688    0  5688    0     0   370k      0 --:--:-- --:--:-- --:--:--  370k
{
    "placement": {
        "cutoverTime": "0",
        "instances": {
            "m3db_local": {
                "endpoint": "127.0.0.1:9000",
                "hostname": "localhost",
                "id": "m3db_local",
                "isolationGroup": "local",
                "port": 9000,
                "shardSetId": 0,
                "shards": [
                    {
                        "cutoffNanos": "0",
                        "cutoverNanos": "0",
                        "id": 0,
                        "sourceId": "",
                        "state": "INITIALIZING"
                    },
                    ...
                    {
                        "cutoffNanos": "0",
                        "cutoverNanos": "0",
                        "id": 63,
                        "sourceId": "",
                        "state": "INITIALIZING"
                    }
                ],
                "weight": 1,
                "zone": "embedded"
            }
        },
        "isMirrored": false,
        "isSharded": true,
        "maxShardSetId": 0,
        "numShards": 64,
        "replicaFactor": 1
    },
    "version": 1
}
```
After a while, the other shards appear to become AVAILABLE:
```
curl  localhost:7201/api/v1/placement | python -m json.tool  
100  5496    0  5496    0     0   383k      0 --:--:-- --:--:-- --:--:--  383k
{
    "placement": {
        "cutoverTime": "0",
        "instances": {
            "m3db_local": {
                "endpoint": "127.0.0.1:9000",
                "hostname": "localhost",
                "id": "m3db_local",
                "isolationGroup": "local",
                "port": 9000,
                "shardSetId": 0,
                "shards": [
                    {
                        "cutoffNanos": "0",
                        "cutoverNanos": "0",
                        "id": 0,
                        "sourceId": "",
                        "state": "AVAILABLE"
                    },
                    ...
                    {
                        "cutoffNanos": "0",
                        "cutoverNanos": "0",
                        "id": 63,
                        "sourceId": "",
                        "state": "AVAILABLE"
                    }
                ],
                "weight": 1,
                "zone": "embedded"
            }
        },
        "isMirrored": false,
        "isSharded": true,
        "maxShardSetId": 0,
        "numShards": 64,
        "replicaFactor": 1
    },
    "version": 2
}
```

It seems the time required for initializing increases with the retention. In the case of 17520H (on a 4G RAM / 100G HDD) machine:
```
Jun 13 18:52:51 m3db1 m3dbnode[17712]: {"level":"info","ts":1560451971.9714265,"msg":"bootstrapped"}
Jun 13 19:07:30 m3db1 m3dbnode[17712]: {"level":"info","ts":1560452850.9808493,"msg":"successfully parsed placement","version":2}
Jun 13 19:07:30 m3db1 m3dbnode[17712]: {"level":"info","ts":1560452850.9817107,"msg":"received update from kv topology watch"}
Jun 13 19:07:30 m3db1 m3dbnode[17712]: {"level":"info","ts":1560452850.9844728,"msg":"bootstrap finished","namespace":"default","duration":0.000015317}
Jun 13 19:07:30 m3db1 m3dbnode[17712]: {"level":"info","ts":1560452850.9846907,"msg":"cluster db successfully marked shards as available","shards":[37,43,18,28,35,42,20,39,46,51,52,4,7,29,58,0,8,30,54,57,62,2,10,26,32,40,63,49,53,55,3,19,24,33,41,38,44,50,9,11,12,22,25,60,15,31,36,56,61,13,17,21,1,16,47,45,6,14,23,34,59,5,27,48]}
```

Also, mem usage is very high (3.5G, on n idling node after shard initialization):
```
# free -mh
              total        used        free      shared  buff/cache   available
Mem:           3.9G        3.5G        127M        636K        238M        135M
Swap:            0B          0B          0B

ps -eo rss,pid,user,command --sort -rss | grep m3dbnode
3513540 17712 root   /usr/bin/m3dbnode -f /etc/m3db/m3dbnode.yaml
```

### CPU / Heap Profiles

The heap dump was generated after all shards were marked as AVAILABLE and m3dbnode receiving 0 traffic:
![profile001](https://user-images.githubusercontent.com/33034/59461392-910b1b80-8e21-11e9-863b-8beb078f168a.png)

**Comment by martin-mao** [comment]
@benraskin92 Here's the long initialization issue for long retention periods. It does complete successfully eventually, but why is it taking so long?

**Comment by richardartoul** [comment]
This takes awhile because M3DB creates data files going as far back as the retention. I.E in your case it creates 2 years worth of files.

17520 (retention hours) / 168 (blocksize M3DB will pick for 2 years of retention I think) * 64 (number of shards) * 7 (number of files to create per namespace/shard/block combination)

We have plans to special case some logic to parallelize and speed up this process in the case where you’re creating a new cluster, but it’s not prioritized right now. It’s also probably especially slow on your machine since you’re not using an SSD and the host probably doesn’t have enough memory to keep all inodes in kernel caches.

If the amount of data that is going to be stored is small (and as a result the cluster doesn’t need to scale to too many nodes) you can just reduce the number of shards which will linearly decrease the amount of files that need to be created on boot.

That said I probably wouldn’t optimize the cluster for this situation alone since it’s a one time cost.

Also are you planning on running with R.F=1 in production?

**Comment by Freeaqingme** [comment]
> Also are you planning on running with R.F=1 in production?

Yes. I fully realize it's not suggested, but right now it's all based off a single hypervisor so there's no point in deploying a distributed setup. M3 is overkill for my use case, but wanted to get some hands-on experience with it for a later project I've got planned.

> That said I probably wouldn’t optimize the cluster for this situation alone since it’s a one time cost.

I agree that if it's a one-time thing it shouldn't deserve too much effort. However, it is one of the first steps new users have with M3, and therefore it can't hurt if this part of the process is nicely polished as well :) Insofar there's a problem here, it's that initially I thought that everything was simply stuck at INITIALIZING. The logfile says "bootstrapped", and from there on out everything is quiet.

I think that if a message would be printed every 30 secs or so with the progress of initialization, the user would know what's going on and what they are waiting for.

**Comment by Freeaqingme** [comment]
Also, is it required for all shards to be in state INITIALIZING while the files for the entire retention period are created?

I just created a namespace with a 168h retention. After a few minutes all shards are marked as AVAILABLE. Afterwards, I create a second namespace with a 17520h retention. While the allocation process is running, the shards remain in state AVAILABLE.

**Comment by Freeaqingme** [comment]
To add to this one more time, I don't actually think it's caused by slow IO (HDD's). The m3dbnode process is utilizing 100% of a single core. Inside the 2 core VM,`top` reports 50% "user" cpu load, and 0% "wait". If it were caused by slow IO the "wait" would be way higher (and, presumably, "user" near-zero).

When I run `strace` on the process I can almost only see futex() calls. Can it be that there's some kind of lock contention going on?

**Comment by richardartoul** [comment]
Ok so its a little more complicated than I admitted (good on you for catching the nuance). If we really get into the weeds, the database has some safeguards about marking shards as AVAILABLE and won't do so until it has performed at least one snapshot since it received new shards. In order to perform a snapshot the dbnode will first flush all missing files for its retention (implementation detail). The snapshot requirement is a durability safeguard to ensure that nodes don't mark themselves as "owning" data after topology changes until the data they have streamed from their peers has been made durable.

This is why when you create a new cluster the switch to AVAILABLE gets blocked for awhile bootstrapping the first namespace, but when you add a new namespace it doesn't matter (because you haven't cause a data ownership change there is no need for any shard state change to take place).

There are a couple ways we could solve this. One would be to special case the logic around new clusters, the other would be to allow the cluster to "bootstrap" even before a namespace is added, and there are a couple other ways also (we currently have [this issue](https://github.com/m3db/m3/issues/1563) open with a potential solution, but I'm no longer certain its the best solution. We just haven't really gotten around to them yet / prioritized them. 

The `futex` thing is very interesting though. Do you mind taking a CPU profile of a node while its happening? My guess is its just some locking in the Go runtime since its blocked on I/O but I'd be curious to see whats going on. You can get a CPU profile with the command below:

```
`curl <HOST_NAME>:<PORT(default 7201)>/debug/pprof/profile?seconds=5 > m3coord_cpu.out`
```

**Comment by Freeaqingme** [comment]
> Do you mind taking a CPU profile of a node while its happening?

Sure thing:
[m3coord_cpu.out.txt](https://github.com/m3db/m3/files/3292352/m3coord_cpu.out.txt)

The same pprof stuff also reports 0 contended mutexes btw.

**Comment by richardartoul** [comment]
@Freeaqingme 

So clearly some room for improvement here...

<img width="1558" alt="Screen Shot 2019-06-14 at 6 25 09 PM" src="https://user-images.githubusercontent.com/9171254/59541986-cbcc8c80-8ed1-11e9-9471-04de2e6c1589.png">

Could sworn we had an open issue for this function allocating so much, but I can't find it.

This function really doesn't need to allocate so much, it just needs to be smarter about parsing so I think we can fix this.

**Comment by Freeaqingme** [comment]
Hmz, I set up a namespace with a retention of 17520h and a blocksizeduration of 30m in order to reduce memory usage. I aborted the creation of this namespace after 15 hrs. A reason this process was probably slow, is that there were 40K files in every shard directory.
$ ls -alh /var/lib/m3db/snapshots/metrics-2year/32 | wc -l
40540

Performing actions on a directory with a lot of files will automatically slow things down as well. There could be room for improvement there as well.

**Comment by signmem** [comment]
@Freeaqingme  
Did you test  restart m3db cluste ..?  How long did the namespace startup complete?  
I have tested restart m3db cluster.  
I won't be able to perform the query until five hours later.  Because namespace did not bootstrap finished

**Comment by Freeaqingme** [comment]
I was forced to test that after it was OOM'ed (I presume during block rollover). However, I wasn't able to get it back up because it kept getting OOM'ed right after start-up.

**Comment by theRealWardo** [comment]
I was hoping to use m3db to retain really long term timeseries data, but this issue basically makes m3db not an option for our use case.

I naively attempted to bootstrap a cluster with a 100 year retention thinking "hey let's keep this data forever go on and page me when you start using too much disk and maybe I'll grow it" but then I discovered all 8 cores on my machine utilized at 100% and about 10G of disk space instantly being consumed by m3db trying to bootstrap. it feels very unnecessary to consume all these resources to just bootstrap the system. is there a proposal for how to fix this or will this forever be an issue for the current architecture of the system?

@robskillington we were considering https://chronosphere.io/ as well but wanted to try the open source project first - does the hosted product have the same issue with long retention? happy to provide more information if it is helpful.
